src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -214,29 +166,75 @@ export class MeetingManager implements AudioVideoObserver {
     this.audioVideoObservers = {};
   }
 
-  async join({
-    meetingInfo,
-    attendeeInfo,
-    deviceLabels = DeviceLabels.AudioAndVideo,
-    eventReporter,
-    meetingManagerConfig,
-  }: MeetingJoinData) {
-    if (meetingManagerConfig) {
-      this.meetingManagerConfig = meetingManagerConfig;
-    }
-    this.configuration = new MeetingSessionConfiguration(
-      meetingInfo,
-      attendeeInfo
+  async join(meetingJoinData: MeetingJoinData): Promise<void> {
+    const { meetingSessionConfiguration } = meetingJoinData;
+    const {
+      deviceLabels,
+      eventReporter,
+      logger,
+      enableWebAudio,
+      activeSpeakerPolicy,
+    } = this.parseMeetingManagerJoinData(meetingJoinData);
+    this.meetingSessionConfiguration = meetingSessionConfiguration;
+    this.meetingId = this.meetingSessionConfiguration.meetingId;
+    this.logger = logger;
+
+    const deviceController = new DefaultDeviceController(logger, {
+      enableWebAudio: enableWebAudio,
+    });
+    this.meetingSession = new DefaultMeetingSession(
+      meetingSessionConfiguration,
+      logger,
+      deviceController,
+      eventReporter
     );
 
-    this.meetingRegion = meetingInfo.MediaRegion;
-    this.meetingId = this.configuration.meetingId;
+    this.audioVideo = this.meetingSession.audioVideo;
 
-    if (eventReporter) {
-      this.eventReporter = eventReporter;
-    }
+    // When an attendee leaves, we remove AudioVideoObservers and nullify AudioVideoFacade and MeetingSession object.
+    // This results into missing few meeting events triggered with audioVideoDidStop such as meetingEnded, meetingFailed
+    // and meetingStartFailed. We may also loose audioInputUnselected and videoInputUnselected events as we choose null
+    // devices when an attendee leaves. When a new AudioVideoFacade object is created remove and re-add the
+    // eventDidReceive observer which wont leak.
+    this.audioVideo.removeObserver(this.eventDidReceiveRef);
+    this.audioVideo.addObserver(this.eventDidReceiveRef);
 
-    await this.initializeMeetingSession(this.configuration, deviceLabels);
+    this.setupAudioVideoObservers();
+    this.setupDeviceLabelTrigger(deviceLabels);
+    await this.listAndSelectDevices(deviceLabels);
+    this.publishAudioVideo();
+    this.setupActiveSpeakerDetection(activeSpeakerPolicy);
+    this.meetingStatus = MeetingStatus.Loading;
+    this.publishMeetingStatus();
+  }
+
+  private parseMeetingManagerJoinData(
+    meetingJoinData: MeetingJoinData
+  ): MeetingManagerJoinOptions {
+    const deviceLabels =
+      meetingJoinData?.options?.deviceLabels ?? DeviceLabels.AudioAndVideo;
+    const eventReporter = meetingJoinData?.options?.eventReporter;
+    const logLevel = meetingJoinData.options?.logLevel ?? LogLevel.WARN;
+    const postLoggerConfig = meetingJoinData?.options?.postLoggerConfig;
+    const logger =
+      meetingJoinData?.options?.logger ??
```

Comment:
I think this should come first as if logger is set, logLevel and postLoggerConfig can be ignored. Just an old fashion if